### PR TITLE
fix game mode error after portal to different game mode world

### PR DESCRIPTION
**Pull request by changyongGuo**
update game mode and capability after portal to another world.
this fix issue #4207

after fix:

creative mode in nether world
creative in nether

survival mode in normal world
survival in normal world

Best regards
Changyong

### CODE REVIEW

File: src/Entities/Player.cpp [hunk 1]
```diff
@@ -2020,6 +2029,9 @@ bool cPlayer::DoMoveToWorld(cWorld * a_World, bool a_ShouldSendRespawn, Vector3d
 			{
 				ch->SendWeather(a_World->GetWeather());
 			}
+
+			// Update game mode
+			SetGameMode(a_World->GetGameMode());
```
Comment (NiLSPACE) [inline src/Entities/Player.cpp]: Perhaps this should be an option in the world.ini? Something like `ForceGameModeOnEnterWorld` or something.

**Comment by peterbell10** [comment]
I'm not sure this is the right solution, it seems like it would be annoying if players had to reset their gamemode every time they change worlds.  Also, it looks like inheriting the gamemode from the world is only supposed to happen when in `gmNotSet`:
https://github.com/cuberite/cuberite/blob/06eb5cc70da76f6a95afa9c9a3d692ca39e39a76/src/Entities/Player.cpp#L1212-L1216

I think a better solution might be to keep the player in `gmNotSet` until their gamemode is explicitly set by a plugin. In which case this would need to be changed:
https://github.com/cuberite/cuberite/blob/2d3013d1d35a2bcba63cffd9012898defff26f93/src/ClientHandle.cpp#L385-L388
And possibly other places in the code as well.

**Comment by changyongGuo** [comment]
@peterbell10 
I agree, 
1, game mode will inherit from world game mode if current player game mode is gmNotSet.
2, game mode will be not changed if game mode is already set
3, remove ForceGameModeOnEnterWorld from option list

I will submit a new commit later

**Comment by changyongGuo** [comment]
new commit submitted, now player game mode can correctly inherit from world after portal to other world.

File: src/Entities/Player.cpp [hunk 2]
```diff
@@ -2023,11 +2050,12 @@ bool cPlayer::DoMoveToWorld(cWorld * a_World, bool a_ShouldSendRespawn, Vector3d
 		}
 
 		// Broadcast the player into the new world.
+		a_World->BroadcastPlayerListAddPlayer(*this);
```
Comment (peterbell10) [inline src/Entities/Player.cpp]: I don't think this is right, the player should already be on the list.  Would `BroadcastPlayerListUpdateGameMode` be enough?

**Comment by peterbell10** [comment]
To undo your change to the core plugin you should navigate to Server/Plugins/Core and run 

```
git checkout 842accabb13d626364e558a34aa5922dbea4efd5
```

Then navigate back to the cuberite folder and stage and commit the submodule update as normal.

File: Server/Plugins/APIDump/Classes/World.lua [hunk 3]
```diff
@@ -2144,6 +2144,16 @@ function OnAllChunksAvailable()</pre> All return values from the callbacks are i
 				},
 				Notes = "Returns whether the configuration has DeepSnow enabled.",
 			},
+			IsForceGameModeOnEnterWorld =
```
[the remark this answers is not in the document]
Comment (changyongGuo) [reply Server/Plugins/APIDump/Classes/World.lua]: sorry I forgot this, I will remove it

File: src/Entities/Player.cpp (hunk 2 above)
Comment (changyongGuo) [reply src/Entities/Player.cpp]: After test, BroadcastPlayerListAddPlayer or BroadcastPlayerListUpdateGameMode is not needed when move player to other world.

I had some misunderstand about protocol wiki...

File: src/Entities/Player.cpp [hunk 4]
```diff
@@ -1694,6 +1695,30 @@ Vector3d cPlayer::GetThrowSpeed(double a_SpeedCoeff) const
 
 
 
+
+eGameMode cPlayer::GetEffectiveGameMode(void) const
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/Entities/Player.cpp]: I don't think you're right about the old version, the ternary expression handles that case:

```cpp
eGameMode GetEffectiveGameMode(void) const
{
    return (m_GameMode == gmNotSet) ? m_World->GetGameMode() : m_GameMode;
}
```

[the remark this answers is not in the document]
Comment (changyongGuo) [reply src/Entities/Player.cpp]: sorry...I mis-understand about this.
I thought it use cPlayer::GetGameMode when m_GameMode == gmNotSet.
I will revert modification about this

**Comment by tigerw** [comment]
Amongst other things, members of the Cuberite subspecies are known for their almost obsessive nitpicking. While this has so far helped them fend off competitors by having a cleaner codebase, it remains to be seen whether now, against the backdrop of a changing Minecraft climate and lack of new developers to join their pack, this will prove to be their biggest weakness.

**Comment by changyongGuo** [comment]
@peterbell10 , thanks for your testing, sorry I didn't test much for last commit

**Pull request merged**